src/netconf_acm.c

```diff
@@ -0,0 +1,1862 @@
+        }
+    }
+
+cleanup:
+    /* NACM UNLOCK */
+    pthread_mutex_unlock(&nacm.lock);
+    sr_nacm_free_groups(groups, group_count);
+
+    return err_info;
+}
+
+/**
+ * @brief Check whether diff node siblings can be applied by a user, recursively with children.
+ *
+ * @param[in] diff First diff sibling.
+ * @param[in] user User for the NACM check.
+ * @param[in] parent_op Inherited parent operation.
+ * @param[in] groups Array of collected groups.
+ * @param[in] group_count Number of @p groups.
+ * @param[out] denied_node NULL if access allowed, otherwise the denied access data node.
+ * @return errinfo, NULL on success.
+ */
+static sr_error_info_t *
+sr_nacm_check_diff_r(const struct lyd_node *diff, const char *user, const char *parent_op, char **groups, uint32_t group_count, const struct lyd_node **denied_node)
+{
+    sr_error_info_t *err_info = NULL;
+    const char *op;
+    struct lyd_meta *meta;
+    uint8_t oper;
+    enum sr_nacm_access access;
+
+    LY_LIST_FOR(diff, diff) {
+        /* find operation */
+        LY_LIST_FOR(diff->meta, meta) {
+            if (!strcmp(meta->name, "operation")) {
+                assert(!strcmp(meta->annotation->module->name, "yang"));
+                break;
+            }
+        }
+        if (meta) {
+            op = lyd_get_meta_value(meta);
+        } else {
+            op = parent_op;
+        }
+        assert(op);
+
+        /* get required access operation */
+        switch (op[0]) {
+        case 'n':
+            /* "none" */
+            oper = 0;
+            break;
+        case 'r':
+            /* "replace" */
+            assert(!strcmp(op, "replace"));
+            oper = SR_NACM_OP_UPDATE;
+            break;
+        case 'c':
+            /* "create" */
+            oper = SR_NACM_OP_CREATE;
+            break;
+        case 'd':
+            /* "delete" */
+            oper = SR_NACM_OP_DELETE;
+            break;
+        default:
+            SR_ERRINFO_INT(&err_info);
+            return NULL;
+        }
+
+        /* check access for the node, none operation is always allowed, and partial access is relevant only for
+         * read operation */
+        if (oper) {
+            err_info = sr_nacm_allowed_node(diff, NULL, NULL, oper, groups, group_count, &access);
+            if (err_info) {
+                return err_info;
+            }
+
+            if (!SR_NACM_ACCESS_IS_NODE_PERMIT(access)) {
+                *denied_node = diff;
+            }
+            break;
+        }
+
+        /* go recursively */
+        if (lyd_child(diff)) {
+            err_info = sr_nacm_check_diff_r(lyd_child(diff), user, op, groups, group_count, denied_node);
+            if (err_info) {
+                return err_info;
+            }
+        }
+    }
+
+    return err_info;
+}
+
+API sr_error_info_t *
+sr_nacm_check_diff(const struct lyd_node *diff, const char *user, const struct lyd_node **denied_node)
```

Comment:
As we talked about, please replace functionality with `sr_nacm_apply_changes()` with a new parameter `user`.

Reply:
Will do